src/rdpc/analysesProcessor.ts

```diff
@@ -0,0 +1,281 @@
+ * Creates inputAnalyses-run map for each donor by grouping runs by inputAnalyses,
+ * creates donor-inputAnalyses-run map by determining the latest run for each inputAnalyses,
+ * aggregates donor-inputAnalyses-run map by donor id.
+ * @param analyses RDPC analysis array
+ */
+export const toDonorCentric = (analyses: Analysis[]): DonorDocMap => {
+  const result = analyses.reduce<DonorDocMap>((acc, analysis) => {
+    const donorWithLatestRun = analysis.donors.reduce<DonorDocMap>(
+      (_acc, donor) => {
+        const inputAnalysesMap = _(analysis.runs)
+          .groupBy((run) => {
+            return (
+              HashCode(run.inputAnalyses[0].analysisId) +
+              HashCode(run.inputAnalyses[1]?.analysisId)
+            );
+          })
+          .value() as InputAnalysesRunMap;
+
+        Object.entries(inputAnalysesMap).forEach(([inputId, runs]) => {
+          const latestRun = getLatestRun(runs);
+          const run = latestRun === undefined ? [] : [latestRun];
+          const existingMap = _acc[donor.donorId];
+          _acc[donor.donorId] = {
+            ...existingMap,
+            [inputId]: run,
+          };
+        });
+
+        return _acc;
+      },
+      {}
+    );
+
+    // merge donor-inputAnalyses-run map by donorId, in case same donors appear under multiple analyses
+    Object.entries(donorWithLatestRun).forEach(
+      ([donorId, inputAnalysesMap]) => {
+        const existingMap = acc[donorId] ? acc[donorId] : {};
+        acc[donorId] = {
+          ...inputAnalysesMap,
+          ...existingMap,
+        };
+      }
+    );
+
+    return acc;
+  }, {});
+
+  return result;
+};
+
+const getLatestRun = (runs: Run[]): Run | undefined => {
+  return _(runs)
+    .sortBy(
+      (run) =>
+        (({ COMPLETE: 1, RUNNING: 2, EXECUTOR_ERROR: 3 } as {
+          [k: string]: number;
+        })[run.state])
+    )
+    .head();
+};
+
+export const mergeDonorMaps = (
+  mergedMap: DonorDocMap,
+  toMerge: DonorDocMap
+): DonorDocMap => {
+  Object.entries(toMerge).forEach(([donorId, inputAnalysesMap]) => {
+    const existingMap = mergedMap[donorId] ? mergedMap[donorId] : {};
+    const mergedInputAnalysesMap = {
+      ...existingMap,
+      ...inputAnalysesMap,
+    };
+    mergedMap[donorId] = { ...mergedInputAnalysesMap };
+  });
+
+  return mergedMap;
+};
+
+export const getAllMergedDonor = async (
+  studyId: string,
+  url: string,
+  config?: {
+    chunkSize?: number;
+    state?: StreamState;
+  }
+): Promise<DonorDocMap> => {
+  const stream = analysisStream(studyId, url, config);
+  let mergedDonorsWithAlignmentRuns: DonorDocMap = {};
+
+  for await (const page of stream) {
+    logger.info(`Streaming ${page.length} sequencing experiment analyses...`);
+    const donorPerPage = toDonorCentric(page);
+    mergedDonorsWithAlignmentRuns = mergeDonorMaps(
+      mergedDonorsWithAlignmentRuns,
+      donorPerPage
+    );
```

Comment:
since `mergeDonorMaps` mutates `mergedDonorsWithAlignmentRuns`.

`mergeDonorMaps(...)` isn't a pure expression so I think it reads better to just write it as a statement without the `mergedDonorsWithAlignmentRuns = ` so it's clearer that there's a side effect.

Also just make `mergedDonorsWithAlignmentRuns` a `const` cause there's no reassignment :)